test/performance/benchmarks/scale-from-zero/continuous/main.go

```diff
@@ -150,37 +179,68 @@ func parallelScaleFromZero(ctx context.Context, clients *test.Clients, objs []*v
 	wg.Wait()
 }
 
-func runScaleFromZero(ctx context.Context, clients *test.Clients, idx int, ro *v1a1test.ResourceObjects) (time.Duration, error) {
-	log.Printf("%02d: waiting for deployment to scale to zero.", idx)
+func runScaleFromZero(ctx context.Context, clients *test.Clients, idx int, ro *v1a1test.ResourceObjects) (
+	time.Duration, time.Duration, error) {
 	selector := labels.SelectorFromSet(labels.Set{
 		serving.ServiceLabelKey: ro.Service.Name,
 	})
 
-	if err := performance.WaitForScaleToZero(ctx, testNamespace, selector, 2*time.Minute); err != nil {
-		m := fmt.Sprintf("%02d: failed waiting for deployment to scale to zero: %v", idx, err)
+	watcher, err := clients.KubeClient.Kube.AppsV1().Deployments(testNamespace).Watch(
+		metav1.ListOptions{LabelSelector:selector.String()})
+	if err != nil {
+		m := fmt.Sprintf("%02d: unable to watch the deployment for the service: %v", idx, err)
 		log.Println(m)
-		return 0, errors.New(m)
+		return 0, 0, errors.New(m)
 	}
+	defer watcher.Stop()
+
+	ddch := watcher.ResultChan()
+	sdch := make(chan struct{})
+	errch := make(chan error)
 
+	go func() {
+		log.Printf("%02d: waiting for endpoint to serve request", idx)
+		url := ro.Route.Status.URL.URL()
+		_, err := pkgTest.WaitForEndpointStateWithTimeout(
+			clients.KubeClient,
+			log.Printf,
+			url,
+			pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(helloWorldExpectedOutput)),
+			"HelloWorldServesText",
+			test.ServingFlags.ResolvableDomain, waitToServe,
+		)
+		if err != nil {
+			m := fmt.Sprintf("%02d: the endpoint for Route %q at %q didn't serve the expected text %q: %v", idx, ro.Route.Name, url, helloWorldExpectedOutput, err)
+			log.Println(m)
+			errch <- errors.New(m)
+		} else {
+			sdch <- struct{}{}
+		}
+	}()
+
+	// Get the duration to serve and the duration that deployment spec is changed.
+	var sd, dd time.Duration
 	start := time.Now()
-	log.Printf("%02d: waiting for endpoint to serve request", idx)
-	url := ro.Route.Status.URL.URL()
-	if _, err := pkgTest.WaitForEndpointStateWithTimeout(
-		clients.KubeClient,
-		log.Printf,
-		url,
-		pkgTest.MatchesAllOf(pkgTest.IsStatusOK, pkgTest.MatchesBody(helloWorldExpectedOutput)),
-		"HelloWorldServesText",
-		test.ServingFlags.ResolvableDomain, waitToServe,
-	); err != nil {
-		m := fmt.Sprintf("%02d: the endpoint for Route %q at %q didn't serve the expected text %q: %v", idx, ro.Route.Name, url, helloWorldExpectedOutput, err)
-		log.Println(m)
-		return 0, errors.New(m)
+LOOP:
+	for {
+		select {
+		case event := <-ddch:
```

Comment:
You don't really need the result channel, do you? You can just compute that in the function above.

Reply:
Do you mean the `WaitForEndpointStateWithTimeout` function? That function polls the endpoint by using a ticker and the request interval is 1 sec. If we check deployment spec change in it, the result won't be accurate .

Reply:
Why? Just doing time.Since() in the place of triggering the channel

Reply:
Hmm sorry can you be more specific about "the place of triggering the channel"?  I'm getting a bit confused as it seems the two comments seem to be different things..